File: web/schemas/admin/broadcast.py
from typing import Literal, Optional, List
from pydantic import BaseModel


class ButtonItem(BaseModel):
    text: str
    url: str
    color: Literal["", "danger", "success", "primary"] = ""
    row: int = 0  # buttons with the same row index are placed in one keyboard row


class BroadcastRequest(BaseModel):
    text: str
    filter: Literal["all", "active", "inactive"] = "all"
    buttons: List[ButtonItem] = []


class BroadcastStartResponse(BaseModel):
    broadcast_id: str
    status: str = "queued"


class BroadcastStatusResponse(BaseModel):
    broadcast_id: str
    status: str  # pending | running | completed | failed
    total: int = 0
    sent: int = 0
    failed: int = 0
    error: Optional[str] = None

File: frontend/src/api/admin/broadcast.ts
import { api } from '@/api'

export type BroadcastFilter = 'all' | 'active' | 'inactive'

export interface BroadcastStartResponse {
  broadcast_id: string
  status: string
}

export interface BroadcastStatusResponse {
  broadcast_id: string
  status: 'pending' | 'running' | 'completed' | 'failed'
  total: number
  sent: number
  failed: number
  error?: string
}

export async function startBroadcast(
  text: string,
  filter: BroadcastFilter = 'all',
): Promise<BroadcastStartResponse> {
  const res = await api.post('/admin/broadcast', { text, filter })
  return res.data
}

export async function getBroadcastStatus(
  broadcastId: string,
): Promise<BroadcastStatusResponse> {
  const res = await api.get(`/admin/broadcast/status/${broadcastId}`)
  return res.data
}
